### PR TITLE
Cognito: Added Identity Pool Roles Attachment

### PR DESCRIPTION
**Pull request by Ninir**
## Description
This adds the AWS Cognito Identity Roles association resource, based on the following API: http://docs.aws.amazon.com/cognitoidentity/latest/APIReference/API_SetIdentityPoolRoles.html

(Migrated from https://github.com/hashicorp/terraform/pull/12846)

## Related issues
* https://github.com/hashicorp/terraform/issues/8309

## Tests
```bash
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSCognitoIdentityPoolRolesAttachment_'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/04/21 20:27:48 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSCognitoIdentityPoolRolesAttachment_ -timeout 120m
=== RUN   TestAccAWSCognitoIdentityPoolRolesAttachment_basic
--- PASS: TestAccAWSCognitoIdentityPoolRolesAttachment_basic (59.64s)
=== RUN   TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappings
--- PASS: TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappings (103.70s)
=== RUN   TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappingsWithError
--- PASS: TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappingsWithError (17.80s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/aws    181.171s
```

## TODOs
- [x] Add the Cognito Identity Roles Resource
- [x] Add Cognito Identity Roles Resource validation
- [x] Add the Cognito Identity Roles Resource documentation
- [x] Add Cognito Identity Roles Resource test cases
- [x] Compile & play with it, in order to ensure every cases are handled & guards are present

### CODE REVIEW

File: aws/resource_aws_cognito_identity_pool_roles_attachment.go [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+									"value": {
+										Type:     schema.TypeSet,
+										MaxItems: 1,
+										Required: true,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"ambiguous_role_resolution": {
+													Type:         schema.TypeString,
+													ValidateFunc: validateCognitoRoleMappingsAmbiguousRoleResolution,
+													Optional:     true, // Required if Type equals Token or Rules.
```
Comment (Ninir) [inline aws/resource_aws_cognito_identity_pool_roles_attachment.go]: At the moment, as `token` is required and has only 2 possibles values (Token or Rules), this seems required too. 
The API states that this is [Optional](http://docs.aws.amazon.com/cognitoidentity/latest/APIReference/API_RoleMapping.html#CognitoIdentity-Type-RoleMapping-AmbiguousRoleResolution) though.

**Comment by Ninir** [comment]
@radeksimko This is a much better approach, thank you for the suggestion :)

While tests are passing, I would like to check another time that everything is ok on my side before merging the work:

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappings'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappings -timeout 120m
=== RUN   TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappings
--- PASS: TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappings (47.47s)
=== RUN   TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappingsWithAmbiguousRoleResolutionError
--- PASS: TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappingsWithAmbiguousRoleResolutionError (11.74s)
=== RUN   TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappingsWithRulesTypeError
--- PASS: TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappingsWithRulesTypeError (11.12s)
=== RUN   TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappingsWithTokenTypeError
--- PASS: TestAccAWSCognitoIdentityPoolRolesAttachment_roleMappingsWithTokenTypeError (11.02s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	81.410s
```

File: aws/resource_aws_cognito_identity_pool_roles_attachment_test.go [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+				return nil
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+const baseAWSCognitoIdentityPoolRolesAttachmentConfig = `
```
[the remark this answers is not in the document]
Comment (Ninir) [reply aws/resource_aws_cognito_identity_pool_roles_attachment_test.go]: Isn't it the case as this constant is concatenated to another string but wrapped into a fmt.Sprintf line https://github.com/terraform-providers/terraform-provider-aws/pull/863/files/4716be981696dcd13458a0e660bfe19b1ab15860#diff-213dd3d721447149c2fa3d3ddfb5e5b1R305 ?

[the remark this answers is not in the document]
Comment (radeksimko) [reply aws/resource_aws_cognito_identity_pool_roles_attachment_test.go]: The link isn't valid anymore, but yeah - I thought you could do something like this:

```go
func testAccAWSCognitoIdentityPoolRolesAttachmentConfig_basic(name string) string {
    return fmt.Sprintf(baseAWSCognitoIdentityPoolRolesAttachmentConfig(name)+`
resource "aws_cognito_identity_pool_roles_attachment" "main" {
```

the benefit you get there is you may change `baseAWSCognitoIdentityPoolRolesAttachmentConfig` template and add/remove modifiers as necessary without having to refactor all references.

It's not a big deal though - so feel free to ignore.

[the remark this answers is not in the document]
Comment (Ninir) [reply aws/resource_aws_cognito_identity_pool_roles_attachment_test.go]: Done :) thanks for the advice!

**Comment by Ninir** [comment]
@radeksimko Ready to go on my side :)

**Pull request merged**